onnxruntime/core/providers/cpu/quantization/quantize_linear.cc

```diff
@@ -175,6 +175,13 @@ Status DequantizeLinear<T>::Compute(OpKernelContext* ctx) const {
   return Status::OK();
 }
 
+// Explicit template instantiation is necessary in order to define 16-bit versions of DequantizeLinear
+// without having to use the ONNX registration macros. The 16-bit DequantizeLinear implementations
+// are currently used by the contrib op com.microsoft.DequantizeLinear.
+// TODO(adrianlizarraga): These can be removed when/if the ONNX spec adds support for 16-bit DequantizeLinear.
+template class DequantizeLinear<int16_t>;
+template class DequantizeLinear<uint16_t>;
```

Comment:
@skottmckay I have now removed these explicit template instantiations. The key was _moving_ the `ONNX_CPU_OPERATOR_TYPED_MS_KERNEL` macro calls into this quantize_linear.cc file. This is currently what is done for QLinearConv, which also has both ONNX and MSDomain versions of the operator. I'll do my best to explain what was happening (although you probably already know):

The `quantize_linear.h` file contained templated class declarations for the `DequantizeLinear` and `QuantizeLinear` OpKernels. However, these templated class declarations did not provide an implementation of the `Compute` function.

```C++
template <typename T>
class DequantizeLinear final : public OpKernel {
 public:
  DequantizeLinear(const OpKernelInfo& info) : OpKernel(info) {
  // ...
  }

  Status Compute(OpKernelContext* context) const override;  // !!
};
```

As you know, templated classes are _typically_ fully specified in their header files so that the compiler is able to generate concrete implementations of templated member functions when it encounters a use (e.g., DequantizeLinear<int16_t>) in a translation unit that includes the header file. Because the implementation of the `Compute` functions is embedded in a separate `quantize_linear.cc` file, the compiler was unable to generate int16 versions of the OpKernels for usage within the `ONNX_CPU_OPERATOR_TYPED_MS_KERNEL` macro calls, which occur in a different translation unit (`quantize_ops.cc`). This resulted in linker errors. There were two possible solutions:

1. Do explicit template instantiation in the `quantize_linear.cc` file to force the compiler to generate 16bit implementations of `Compute`. This is what I originally did.
2. Move the `ONNX_CPU_OPERATOR_TYPED_MS_KERNEL` macro calls into the `quantize_linear.cc` file. These macros can now "see" the `Compute` function's definition and thus the compiler can generate 16bit versions. This is what I have now done (and what is done for QLinearConv)

Reply:
Additionally, because the `ONNX_CPU_OPERATOR_TYPED_MS_KERNEL` macro calls are guarded by `#ifndef DISABLE_CONTRIB_OPS`, the 16-bit implementations will only be generated when contrib ops are enabled in the build. The previous explicit template instantiation approach always forced the 16bit versions to be generated, which raised valid binary-size concerns.